enterprise/kernel/src/main/java/org/neo4j/kernel/impl/enterprise/lock/forseti/ForsetiClient.java

```diff
@@ -770,10 +773,12 @@ private void clearWaitList()
         waitList.put( clientId );
     }
 
-    private void markAsWaitingFor( ForsetiLockManager.Lock lock, ResourceType type, long resourceId )
+    private void waitFor( ForsetiLockManager.Lock lock, ResourceType type, long resourceId, int tries )
     {
+        waitingForLock = lock;
         clearWaitList();
         lock.copyHolderWaitListsInto( waitList );
+        applyWaitStrategy( type, tries );
```

Comment:
For some reason this refactoring makes this clearer than it was before - which is good - but it brings up a comment unrelated to the actual change you have made:

We are applying the wait strategy _before_ we check for deadlocks, which means that in the case of a deadlock we will spend extra time (waiting) before we detect the deadlock. That seems like a bad thing.

There is actually an aspect of this that relates to the change at hand: since the lock acquisition retry loop _surrounds_ the call to `waitFor(...)` and the implementation clears the `waitingForLock` field on exit, there is a possibility that the deadlock detection algorithm will see null in the `waitingForLock` field of a LockClient that is actually trying to wait for the lock, thus deferring the detection of the deadlock (in theory indefinitely, in reality probably not). If we instead assigned `waitingForLock` in the loop we would not have to clear it in between, but could just re-set it to the newly discovered lock in each iteration, and then clear it afterwards when we know we are not waiting for any lock.